contracts/managers/allocation-pricers/BinaryAllocationPricer.sol

```diff
@@ -0,0 +1,461 @@
+     * @param  _coreInstance                        The address of the Core contract
+     * @param  _setTokenFactoryAddress              The address of SetTokenFactory used to create
+     *                                              new collateral
+     */
+    constructor(
+        ERC20Detailed _baseAssetInstance,
+        ERC20Detailed _quoteAssetInstance,
+        IOracle _baseAssetOracleInstance,
+        IOracle _quoteAssetOracleInstance,
+        ISetToken _baseAssetCollateralInstance,
+        ISetToken _quoteAssetCollateralInstance,
+        ICore _coreInstance,
+        address _setTokenFactoryAddress
+    )
+        public
+    {
+        // Get components of collateral instances
+        address[] memory baseAssetCollateralComponents = _baseAssetCollateralInstance.getComponents();
+        address[] memory quoteAssetCollateralComponents = _quoteAssetCollateralInstance.getComponents();
+
+        // Make sure collateral instances are using the correct base and quote asset
+        require(
+            baseAssetCollateralComponents[0] == address(_baseAssetInstance),
+            "MACOStrategyManager.constructor: Base collateral component must match base asset."
+        );
+
+        require(
+            quoteAssetCollateralComponents[0] == address(_quoteAssetInstance),
+            "MACOStrategyManager.constructor: Quote collateral component must match quote asset."
+        );
+
+        // Set state frome constructor params
+        baseAssetCollateralInstance = _baseAssetCollateralInstance;
+        quoteAssetCollateralInstance = _quoteAssetCollateralInstance;
+
+        baseAssetInstance = _baseAssetInstance;
+        quoteAssetInstance = _quoteAssetInstance;
+
+        baseAssetOracleInstance = _baseAssetOracleInstance;
+        quoteAssetOracleInstance = _quoteAssetOracleInstance;
+
+        // Query decimals of base and quote assets
+        baseAssetDecimals = _baseAssetInstance.decimals();
+        quoteAssetDecimals = _quoteAssetInstance.decimals();
+
+        // Set Core and setTokenFactory
+        coreInstance = _coreInstance;
+        setTokenFactoryAddress = _setTokenFactoryAddress;
+    }
+
+    /*
+     * Determine the next allocation to rebalance into. If the dollar value of the two collateral sets is more
+     * than 4x different from each other then create a new collateral set. If currently 100% in baseAsset then
+     * a new quote collateral set is created if 0% in baseAsset then a new base collateral set is created.
+     *
+     * @param  _targetBaseAssetAllocation       Target allocation of the base asset
+     * @param  _currentCollateralSet            Instance of current set collateralizing RebalancingSetToken
+     * @return address                          The address of the proposed nextSet
+     * @return uint256                          The USD value of current Set
+     * @return uint256                          The USD value of next Set
+     */
+    function determineNewAllocation(
```

Comment:
Does this function need to be permissioned to only the SetToken it's connected to? 

Ah yes, the risk that you mention is that someone else can arbitrarily change the baseSet / quoteSet instance.

I guess it depends how reusable we see this being for different Sets.

Reply:
The risk is that if multiple Sets use it that have varying performance, we may see new collateral Sets get created often / a lot since the prices are so different.